### PR TITLE
feat: unauthenticated `pull` and interactive auth with `push`

### LINKED ISSUE #539: flox push prompts for login
# Acceptance Criteria

**Given** I am unauthenticated with flox hub
**And** I have an environment in a folder called `material-ui` that contains a single package, `yarn`
**When** I `flox push` from `~/material-ui`
**Then** I see login prompt

_Pushing an environment_
**And When** I create an account and login successfully (`ghudgins` for me with github)
**Then** I see my environment, `material-ui` is now in floxhub
**And** I see a confirmation

### LINKED ISSUE #663: `flox pull` & `flox activate -r` allows anonymous users
# Acceptance Criteria

**Given** I have not `flox auth login`
**When** I `flox activate -r ghudgins/cli`
**Then** I am taken to an interactive shell for the env (today it errors)
**And When** I `flox pull ghudgins/cli` 
**Then** I pull successfully (today it errors)

Note: the API already allows this behavior.

### PR DESCRIPTION
**Pull request by ysndr**
closes #539
closes #663

### CODE REVIEW

**Comment by ysndr** [comment]
@ghudgins for your reference

File: cli/flox/src/commands/environment.rs [hunk 1]
```diff
@@ -783,8 +783,47 @@ pub struct Push {
...
 impl Push {
-    pub async fn handle(self, flox: Flox) -> Result<()> {
+    pub async fn handle(self, mut flox: Flox) -> Result<()> {
         subcommand_metric!("push");
+
+        if flox.floxhub_token.is_none() {
+            let options = indoc! {"
+                * login to floxhub with 'flox auth login',
+                * set the 'floxhub_token' field to '<your token>' in your config
+                * set the '$FLOX_FLOXHUB_TOKEN=<your_token>' environment variable."};
```
Comment (ysndr) [inline cli/flox/src/commands/environment.rs:793]: this makes error messages a bit wordy, but at least tells you what to do:

```
ERROR: You are not logged in to floxhub.
    
    Can not automatically login to floxhub in non-interactive context.
    
    To login you can either
    * login to floxhub with 'flox auth login',
    * set the 'floxhub_token' field to '<your token>' in your config
    * set the '$FLOX_FLOXHUB_TOKEN=<your_token>' environment variable.
```

We can also just refer to the config or auth manpage which might be a bit more concise

File: cli/flox/src/commands/environment.rs [hunk 2]
```diff
@@ -783,8 +783,47 @@ pub struct Push {
...
+                let message = formatdoc! {"
+                    You are not logged in to floxhub.
+
+                    Can not automatically login to floxhub in non-interactive context.
+
+                    To login you can either
+                    {options}
+                "};
+
+                bail!(message);
```
Comment (mkenigs) [inline cli/flox/src/commands/environment.rs:817]: nonblocking: I think we want this to start with `🛑 ERROR`

File: cli/flox-rust-sdk/src/providers/git.rs [hunk 3]
```diff
@@ -466,18 +466,30 @@ impl GitCommandProvider {
...
 
         if force {
             command.arg("--force");
         }
 
-        GitCommandProvider::run_command(&mut command)?;
+        match GitCommandProvider::run_command(&mut command) {
+            Ok(_) => Ok(()),
+            Err(GitCommandError::BadExit(_, _, stderr)) if stderr.contains("DENIED") => {
+                Err(GitRemoteCommandError::AccessDenied)
```
Comment (mkenigs) [inline cli/flox-rust-sdk/src/providers/git.rs:484]: question nonblocking: how come you aren't bubbling the error text up? Because if it contains DENIED you know there won't be any additional helpful information?

File: cli/flox-rust-sdk/src/models/environment/managed_environment.rs [hunk 4]
```diff
@@ -798,12 +806,9 @@ impl ManagedEnvironment {
...
         let temp_floxmeta_path = tempfile::tempdir_in(&flox.temp_dir).unwrap().into_path();
 
-        let token = flox
-            .floxhub_token
-            .as_ref()
-            .ok_or(ManagedEnvironmentError::OpenFloxmeta(
-                FloxmetaV2Error::LoggedOut,
-            ))?;
+        // caller must ensure that a token is present
+        // or pushing git operations will fail
```
Comment (mkenigs) [inline cli/flox-rust-sdk/src/models/environment/managed_environment.rs:810]: suggestion:
Add a comment about authentication to `push_new`:
```
    /// If access to a remote repository requires authentication,
    /// the floxhub token must be set in the flox instance.
    /// The caller is responsible for ensuring that the token is present and valid.
```
and
```suggestion
        // Caller decides whether to set token
```

File: cli/flox/src/commands/environment.rs [hunk 5]
```diff
@@ -783,8 +783,47 @@ pub struct Push {
...
+                * login to floxhub with 'flox auth login',
+                * set the 'floxhub_token' field to '<your token>' in your config
+                * set the '$FLOX_FLOXHUB_TOKEN=<your_token>' environment variable."};
+
+            let dialog_help = formatdoc! {"
+                Alternatively you can also
+                {options}"};
+
+            let dialog = Dialog {
+                message: "You are not logged in to floxhub. Would you like to login now?",
```
Comment (mkenigs) [inline cli/flox/src/commands/environment.rs:800]: suggestion: as logging in is the only possible solution, we don't need to confirm logging in. See style guide

File: cli/tests/environment-pull.bats [hunk 6]
```diff
@@ -84,17 +84,6 @@ function update_dummy_env() {
...
-# bats test_tags=pull:l1
-@test "l1: pull login: running flox pull before user has login metadata prompts the user to login" {
-
-  unset FLOX_FLOXHUB_TOKEN; # logout, effectively
-
-  run "$FLOX_BIN" pull --remote owner/name # dummy remote as we are not actually pulling anything
-  assert_failure
-  assert_output --partial 'Please login to floxhub with `flox auth login`'
-}
-
```
Comment (mkenigs) [inline cli/tests/environment-pull.bats:97]: question blocking: is there coverage for both an authenticated and unauthenticated pull?

File: cli/flox-rust-sdk/src/models/environment/managed_environment.rs [hunk 7]
```diff
@@ -798,12 +806,9 @@ impl ManagedEnvironment {
...
+        let token = flox.floxhub_token.as_deref();
```
Comment (mkenigs) [inline cli/flox-rust-sdk/src/models/environment/managed_environment.rs:811]: suggestion nonblocking: I think a doc comment on flox.floxhub_token would be helpful:
- It's populated from config during flox initialization
- Checking for `None` can be used to check if the use is logged in

**Review by mkenigs — CHANGES_REQUESTED** [review]

File: cli/flox-rust-sdk/src/providers/git.rs (hunk 3 above)
Comment (ysndr) [reply cli/flox-rust-sdk/src/providers/git.rs:487]: but i added `debug` messages to catch the entire thing at least somewhere

Comment (mkenigs) [reply cli/flox-rust-sdk/src/providers/git.rs:487]: Makes sense I was just curious. When you `match` on the error you return a different error rather than wrapping it with `map_err` which is different than what we usually do.

File: cli/flox/src/commands/environment.rs (hunk 5 above)
Comment (ysndr) [reply cli/flox/src/commands/environment.rs:800]: It's still requiring user interaction (the login) asking if users want to do that seems reasonable -- i understand the style guide as avoinding interaction if an automated thing can be done

Comment (mkenigs) [reply cli/flox/src/commands/environment.rs:800]:
```
Example: “`flox push` without auth should automatically flox auth login and print a sentence: “Auth information not found. Redirecting to ‘flox auth login’.”
```

File: cli/tests/environment-pull.bats (hunk 6 above)
Comment (ysndr) [reply cli/tests/environment-pull.bats:75]: pull works regardless of being logged in.
The tests are running with a token set

File: cli/flox-rust-sdk/src/models/environment/managed_environment.rs (hunk 7 above)
Comment (ysndr) [reply cli/flox-rust-sdk/src/models/environment/managed_environment.rs:840]: done

File: cli/tests/environment-pull.bats (hunk 6 above)
Comment (mkenigs) [reply cli/tests/environment-pull.bats:75]: So shouldn't we run a test without a token set?

File: cli/flox/src/commands/environment.rs (hunk 5 above)
Comment (ysndr) [reply cli/flox/src/commands/environment.rs:800]: hmm

File: cli/tests/environment-pull.bats (hunk 6 above)
Comment (ysndr) [reply cli/tests/environment-pull.bats:75]: it doesnt make a difference in these tests anyway as the token is not not used for authentication (we're pulling from a bare repo).
But i can add a test showing that we dont block on it..

File: cli/flox/src/commands/environment.rs (hunk 5 above)
Comment (ysndr) [reply cli/flox/src/commands/environment.rs]: ok it does say redirect automatically 🤔
Changed..

**Review by mkenigs — APPROVED** [review]
I think some of the messages could be tweaked slightly but we can do that after more input from product

**Comment by ysndr** [comment]
had to format the test file to avoid (🤞🏼) conflicts with main

**Review by mkenigs — APPROVED** [review]

**Review by mkenigs — APPROVED** [review]
